app/controllers/keybase_proofs_controller.rb

```diff
@@ -0,0 +1,45 @@
+class KeybaseProofsController < ApplicationController
+  before_action :require_logged_in_user
+  before_action :check_new_params, only: :new
+  before_action :check_user_matches, only: :new
+
+  def new
+    @kb_username = params[:kb_username]
+    @kb_signature = params[:kb_signature]
+    @kb_ua = params[:kb_ua]
+    @kb_avatar = Keybase.avatar_url(@kb_username)
+  end
+
+  def create
+    kb_username = post_params[:kb_username]
+    kb_signature = post_params[:kb_signature]
+    kb_ua = post_params[:kb_ua]
+    if Keybase.proof_valid?(kb_username, kb_signature, @user.username)
+      @user.add_or_update_keybase_proof(kb_username, kb_signature)
+      @user.save!
+      return redirect_to Keybase.success_url(kb_username, kb_signature, kb_ua, @user.username)
+    else
+      flash[:error] = "Failed to connect your account to Keybase. Try again from Keybase."
+      return redirect_to settings_path
+    end
+  end
+
+private
+
+  def check_user_matches
+    unless @user.username.casecmp(params[:username]).zero?
+      flash[:error] = "not logged in as the correct user"
+      return redirect_to settings_path
+    end
+  end
+
+  def post_params
+    params.require(:keybase_proof).permit(:kb_username, :kb_signature, :kb_ua, :username)
+  end
+
+  def check_new_params
+    redirect_to settings_path unless [:kb_username, :kb_signature, :kb_ua, :username].all? do |k|
```

Comment:
```suggestion
    redirect_to settings_path unless [:kb_username, :kb_signature, :kb_ua, :username].all?(&:present?)
```

Could you explain this logic? Why do we push someone back to `/settings` without an error message if all the params are present? Should this be sending them back if any are missing instead of if all of them are present?

Reply:
this redirects to the main settings _unless_ they're all there. also, this should never happen in the normal user flow, since these parameters will all be populated by keybase (there aren't links to the `/new` path anywhere because you need to start the proof flow from keybase). so you'd only hit this redirect if you were manually messing with the url trying to do something weird.